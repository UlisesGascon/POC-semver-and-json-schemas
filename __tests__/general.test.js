const Ajv = require('ajv')
const addFormats = require('ajv-formats')
const semver = require('semver')
const jsonSchemasLibrary = require('../src/index')

const ajv = new Ajv()
addFormats(ajv)

const schemas = Object.keys(jsonSchemasLibrary.user).map(key => {
  return [jsonSchemasLibrary.user[key].name, jsonSchemasLibrary.user[key]]
})

describe('All Schema files validation', () => {
  test.each(schemas)(
    'Should %s contain the VALID schema',
    (name, schema) => {
      expect(() => ajv.compile(schema.schema)).not.toThrow()
    }
  )

  test.each(schemas)(
    'Should %s contain a VALID version semver compatible',
    (name, schema) => {
      const version = semver.valid(schema.version)
      expect(version).not.toEqual(null)
    }
  )

  test.each(schemas)(
    'Should %s contain a VALID semver schema Compatibility',
    (name, schema) => {
      // At least 3 characters (1.x)
      expect(schema.schemaCompatibility.length >= 3).toEqual(true)
    }
  )

  test.each(schemas)(
    'Should %s contain a VALID name',
    (name, schema) => {
      // At least 5 characters
      expect(schema.name.length >= 5).toEqual(true)
    }
  )
})

describe('All Schema files validation', () => {
  it('Should export a valid library to be consumed by other projects', () => {
    expect(jsonSchemasLibrary).toMatchSnapshot()
  })
})
