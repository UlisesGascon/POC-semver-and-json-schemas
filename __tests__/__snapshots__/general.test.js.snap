// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Schema files validation Should export a valid library to be consumed by other projects 1`] = `
{
  "user": {
    "created": {
      "name": "userCreated",
      "schema": {
        "additionalProperties": false,
        "properties": {
          "age": {
            "description": "User age",
            "maximum": 120,
            "minimum": 0,
            "type": "integer",
          },
          "email": {
            "description": "User email",
            "format": "email",
            "type": "string",
          },
          "id": {
            "description": "Unique Id for the user",
            "format": "uuid",
            "type": "string",
          },
          "schemaVersion": {
            "description": "Metadata field to determine the json schema version used in the message",
            "maxLength": 14,
            "minLength": 5,
            "pattern": "^(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)$",
            "type": "string",
          },
          "timestamp": {
            "description": "Metadata field to determine the date of the message",
            "minimum": 0,
            "type": "integer",
          },
        },
        "required": [
          "id",
          "email",
          "schemaVersion",
          "timestamp",
        ],
        "title": "userCreated",
        "type": "object",
      },
      "schemaCompatibility": "1.x",
      "version": "1.0.0",
    },
    "deleted": {
      "name": "userDeleted",
      "schema": {
        "additionalProperties": false,
        "properties": {
          "id": {
            "description": "Unique Id for the user",
            "format": "uuid",
            "type": "string",
          },
          "schemaVersion": {
            "description": "Metadata field to determine the json schema version used in the message",
            "maxLength": 14,
            "minLength": 5,
            "pattern": "^(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)\\.(?:0|[1-9]\\d*)$",
            "type": "string",
          },
          "timestamp": {
            "description": "Metadata field to determine the date of the message",
            "minimum": 0,
            "type": "integer",
          },
        },
        "required": [
          "id",
          "schemaVersion",
          "timestamp",
        ],
        "title": "userDeleted",
        "type": "object",
      },
      "schemaCompatibility": "1.x",
      "version": "1.0.0",
    },
  },
}
`;
